### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by krlvi** [gitbutlerapp/gitbutler#2594]
Tracking windows support

---

> Edit by @Qix- 

Hello everyone, Windows support is live. See https://github.com/gitbutlerapp/gitbutler/issues/2594#issuecomment-2047545213 and https://gitbutler.com for links!

**Comment by francois-joubert** [comment]
Really looking forward trying this out on my win machine! Any chance this would run on WSL in the meantime?

**Comment by anaisbetts** [comment]
@francois-joubert It seems to run on my WSL instance but has DPI scaling issues that makes it unusable:

![image](https://github.com/gitbutlerapp/gitbutler/assets/1396/feb361f5-020b-40d3-b42c-fd5fb58262df)

**Comment by Cerabbite** [comment]
@anaisbetts What kind of monitor do you have? On my Dell XPS 17 monitor it looks fine

![image](https://github.com/gitbutlerapp/gitbutler/assets/98265679/41d6c090-0233-4c1b-bc22-b779635ad93c)

**Comment by KroniK907** [comment]
I installed firefox on my wsl instance and set it as my default browser (using microsoft RDP to get into the settings app for the wsl instance). Once I had a working default browser setup, I was able to complete the sign in process and have all features working via wsl.

**Comment by anaisbetts** [comment]
@Cerabbite This is on my laptop screen @ 200% scaling

**Comment by KroniK907** [comment]
I have created a gist that provides a complete walkthrough for getting GitButler working on WSL: https://gist.github.com/KroniK907/24c6edafa2b34af773dc9466dd7e4e00

**Comment by Cerabbite** [comment]
@KroniK907 Thank you, this was way faster than my way which was compiling it from scratch. Though it is still very unstable for me.

![image](https://github.com/gitbutlerapp/gitbutler/assets/98265679/f18cac7c-18f1-4a4a-80df-c2a900ee2521)
But I'll keep trying to see if I can fix my issues

**Comment by Ronsku** [comment]
Looking forward to Windows support! 🥳 
Are there any time estimations?

**Comment by GIRU-GIRU** [comment]
Another windows user here praying for win support, Unreal Engine dev not a great experience on linux, WSL is abit much effort as well

**Comment by Cerabbite** [comment]
@GIRU-GIRU I got GitButler setup on WSL in a matter of minutes. (I did already have WSL setup)

I used this guide by @KroniK907 
> I have created a gist that provides a complete walkthrough for getting GitButler working on WSL: https://gist.github.com/KroniK907/24c6edafa2b34af773dc9466dd7e4e00

**Comment by KroniK907** [comment]
Looks like according to some comments on my gist that the WSL solution is really only viable if you are hosting your repos in WSL, otherwise the lag of having WSL access the windows file system makes it rather laggy and hard to use at best, and completely times out and is not useable at worst. 

However, if you are able to use WSL as your dev environment it works really nicely, especially since VSCode works nicely with project repos in WSL.

**Comment by KroniK907** [comment]
Updated my gist to add instructions for installing GitButler as an AppImage. The main reason for adding the AppImage install instructions was because the AppImage version can update itself, while the .deb has to be manually updated for each release.

Because the team is adding improvements and releasing new updates quickly, it seemed practical to provide install instructions for a version that did not require as much effort to keep up to date.

**Comment by impoetk** [comment]
I have and use WSL but my repos are windows only (WPF) looking forward for windows support

**Comment by 7ahery** [comment]
Can't wait to test out the windows version 🤞

**Comment by dannyking** [comment]
+1 for Windows build

**Comment by bill876** [comment]
I have a setup where GUI apps from WSL run via VcXSrv on Windows side, one issue I encountered:

The window appears, but it is empty, 
and terminal shows no errors, or this error: 
```
[...] 'GLXBadFBConfig'. 
(Details: serial 172 error_code 163 request_code 149 (GLX) minor_code 21) [...]
```

These changes helped me solve the issue: 
disable the `wgl` setting in vcxsrv, and also remove `LIBGL_ALWAYS_INDIRECT` from my linux environment

**Comment by MystPi** [comment]
When Windows is supported, will it have support for the WSL filesystem? I feel like installing gitbutler in WSL isn't the best solution for WSL devs.

**Comment by KroniK907** [comment]
> I feel like installing gitbutler in WSL isn't the best solution for WSL devs.

As a wsl dev, installing GitButler in WSL has been working great for me. But to your question about file access, any windows app can access the `\\wsl.localhost\` network path to the wsl instances. But that may end up being kinda laggy. We won't know for sure what that performance hit will look like till there is a windows build.

**Comment by interceptor** [comment]
I also really want windows support, cannot wait! :)

**Comment by Southclaws** [comment]
Would love to help get this running natively on Windows, what's currently blocking this?

Native Windows dev so happy to help here!

**Comment by MHHenriksen** [comment]
> Would love to help get this running natively on Windows, what's currently blocking this?
> 
> Native Windows dev so happy to help here!

https://github.com/Byron/gitoxide/issues/1287 mentions this issue and has a few things to attack

**Comment by w5l** [comment]
> > Would love to help get this running natively on Windows, what's currently blocking this?
> > Native Windows dev so happy to help here!
> 
> [Byron/gitoxide#1287](https://github.com/Byron/gitoxide/issues/1287) mentions this issue and has a few things to attack

Which says about Windows support:

> There may be rough edges around program and hook invocations, to be smoothed out of time (probably this is about more than Git though, currently unclear)

Which sounds to me like the thing to pick up here is getting a beta (alpha?) Windows build done, and letting people test it? Then we can find said "rough edges", make tickets for them and get them fixed.

**Comment by Southclaws** [comment]
Modern Windows installations will have ssh installed and available as an .exe and I think a link library too.

The git for windows depends on the install, for example I only install Git itself without any of the extras (I don't want their weird bundled terminal emulator and linuxy libraries, I just want git.exe - so I opt for installing only that, which may omit the ssh installs too since I never use ssh anyway) so a check to see if it's in Path like on any other system would make the most sense along with a helpful error message if not, linking to the relevant install docs.

Is there any way SSH could just be packaged with the install in future, either via static linking or a DLL?

**Comment by w5l** [comment]
> Is there any way SSH could just be packaged with the install in future, either via static linking or a DLL?

One of the nice features of using SSH is the option for key based authentication and using `ssh-agent`. You can then add and remove auth and signing keys to `ssh-agent` on the fly (eg. with a password manager). I think this all falls apart if apps start using different SSH libraries, since the `ssh-agent` is linked to a specific `ssh` install. At least, that is my experience while fighting ssh key auth on a box having SSH bundled with Git, SSH bundled with Windows and an OpenSSH install.

Anyway, the "no native ssh transport yet" seems clear enough and solvable (take `ssh` from path). I'm more interested in the "There may be rough edges around program and hook invocations" bit.

**Comment by MHHenriksen** [comment]
> PS. According to some Microsoft blogs, SSH is available as a Windows Feature for general public from Windows 10 version 1803 (released early 2018) onwards.

That's right! It's trivial to install or check if it's available and installed with [DISM](https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/dism-capabilities-package-servicing-command-line-options?view=windows-11) or its equivalent [Powershell commands](https://learn.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse?tabs=powershell), it's as simple as `Add-WindowsCapability -Online -Name OpenSSH.Client*` (the name includes a version number, hence the trailing wildcard). And if you ask me, Windows Server 2019 or Windows 10 (build 1809, not 1803) is entirely fine to have as a minimum requirement.

But wait, there's more! Microsoft is working on [an updated version on GitHub](https://github.com/PowerShell/Win32-OpenSSH), and ship beta versions which can be [installed with winget](https://github.com/PowerShell/Win32-OpenSSH/wiki/Install-Win32-OpenSSH) (**literally just `winget install "openssh beta"`!**) or downloaded as a .zip or .msi from their GitHub releases (and can therefore also be packaged and shipped with GitButler if desired).

Hope that helps!

**Comment by Qix-** [comment]
Just pushed a small fix that should unlock project creation on Windows in cases where it might have failed before. There are other issues (performance being a big one). There are other issues I'm investigating at the moment, however I'd love to know which ones people are running into still. Builds should be un-broken, too.

```
mkdir gitbutler-ui/build
cargo build
pnpm i
pnpm tauri dev
```

Feel free to comment here.

**Comment by A-Lamia** [comment]
ok so some things i have noticed:

- when fetching from origin/main on trunk even though it has completed the task according to the terminal output, it is stuck on `busy...`.
- gitbutler seems to be reporting changes incorrectly saying i have 69 changes when i have 3 modifications and 3 unwatched files.

I'll keep using it and update if i find anything else.

**Comment by maffin-sa** [comment]
> * when fetching from origin/main on trunk even though it has completed the task according to the terminal output, it is stuck on `busy...`.

I experienced this as well, and I noticed in the DevTools window when this happened I eventually received the error: "Project [GUID] does not have a default target set" even though during setup I selected origin/main as the base branch and the UI reflected that as well.

**Comment by Qix-** [comment]
> when fetching from origin/main on trunk even though it has completed the task according to the terminal output, it is stuck on busy....

I think this is a known issue that affects all versions. Definitely will get addressed at some point.

> when clicking on Trunk to view commits on the branch the images seem to be broken.

You mean e.g. profile picture images? Or do you mean something else?

> gitbutler seems to be reporting changes incorrectly saying i have 69 changes when i have 3 modifications and 3 unwatched files.

Yeah I faced this too, was this a repo you had cloned from WSL by chance, and are accessing via the Windows side of things? Either via `/mnt` or the other direction, from Windows -> WSL via the network share? Or was this Windows through and through?

> unable to create a new branch / dragging dropping files does not work.

Yeah, confirmed this on my end as well. Will look into it.

---

> I experienced this as well, and I noticed in the DevTools window when this happened I eventually received the error: "Project [GUID] does not have a default target set" even though during setup I selected origin/main as the base branch and the UI reflected that as well.

This might be a timing/performance thing, at least on my end you could at least get into the virtual branch view and see one branch (but couldn't create any others). Unless I'm misunderstanding what you're saying.

Thanks for testing things!

**Comment by Megasware128** [comment]
When running commit 7eabc10 in dev mode my virtual branch with about 5 files (1 of them is a LFS binary file) is generating a new name every second

**Comment by A-Lamia** [comment]
> You mean e.g. profile picture images? Or do you mean something else?

correct, profiles pictures and are broken.

>Yeah I faced this too, was this a repo you had cloned from WSL by chance, and are accessing via the Windows side of things? Either via /mnt or the other direction, from Windows -> WSL via the network share? Or was this Windows through and through?

everything here is only on windows, no wsl. I did try it on a completely fresh clone and it was still reporting 66 changes very strange.

**Comment by Qix-** [comment]
Gotcha, thank you for the followup. Very strange indeed. Did `git status` show anything in the CLI?

**Comment by A-Lamia** [comment]
>Gotcha, thank you for the followup. Very strange indeed. Did git status show anything in the CLI that could correspond to the changes GitButler was reporting? Mode changes, etc?

Sorry, nothing, when using `git status` modified repo shows correct changes and added files, clean clone shows nothing changed or modified. seems git butler just says that every file is modified though it is correctly picking up on untracked files.

**Comment by Qix-** [comment]
Interesting, alright thanks for the info!

**Comment by ahaoboy** [comment]
I've compiled it successfully in the msys environment, you can download from [gitbutler-msys](https://github.com/ahaoboy/gitbutler-msys/releases) and try it out, it may be missing some necessary dlls, only zlib1.dll and WebView2Loader.dll are added by default, there are some performance issues and bugs, but it's working on my computer, maybe we should open a new issue to track down the various issues in msys.
![PixPin_2024-03-06_17-39-36](https://github.com/gitbutlerapp/gitbutler/assets/19884146/30c1a338-03d3-4e8e-ab8f-b991fb851775)

**Comment by Qix-** [comment]
Interesting, the notify functionality is working for you. Can you create more vbranches and move hunks between them?

**Comment by ahaoboy** [comment]
> Can you create more vbranches and move hunks between them?

I can't create a branch and commit it, perhaps because the relevant tests don't pass on Win.

![PixPin_2024-03-06_23-41-52](https://github.com/gitbutlerapp/gitbutler/assets/19884146/c688c749-6944-4025-8e1c-45e844723895)
![PixPin_2024-03-06_23-44-08](https://github.com/gitbutlerapp/gitbutler/assets/19884146/64921c18-0530-42f3-ba09-9ce62dff45a6)

When I create a branch by clicking the new branch button or drop a file, the console doesn't report any errors, but the UI doesn't change.
```
2024-03-06T15:47:46.432540Z  INFO create_virtual_branch: gitbutler-app\src\virtual_branches\commands.rs:109: close time.busy=19.3ms time.idle=30.5µs project_id="f0122fcb-255e-4b19-adbf-34346860e851" branch=BranchCreateRequest { name: None, ownership: None, order: None, selected_for_changes: None }
2024-03-06T15:47:46.482670Z  INFO can_apply_virtual_branch: gitbutler-app\src\virtual_branches\commands.rs:389: close time.busy=4.44ms time.idle=16.5µs project_id="f0122fcb-255e-4b19-adbf-34346860e851" branch_id="f7e52d17-2571-4237-870a-1aaca6c3dfea"
2024-03-06T15:47:46.483279Z  INFO can_apply_virtual_branch: gitbutler-app\src\virtual_branches\commands.rs:389: close time.busy=4.97ms time.idle=12.3µs project_id="f0122fcb-255e-4b19-adbf-34346860e851" branch_id="f7e52d17-2571-4237-870a-1aaca6c3dfea"
```

**Comment by A-Lamia** [comment]
Ok, tested out recent changes awesome job so far! Currently i can interact with GitButler with mostly no issues:

- Correct git changes being reported.
- Creating virtual branches.
- Stuck git pull spinner.
- Drag and dropping changes to new branch.
- Committing changes.

What is still broken:

- Git push fails.
- Images still fail to load.
- Git commits are credited to GitButler regardless of setting that controls who is the author.
- Can not sign in, get an error saying there was an issue, error in DevTools shows (Unauthorized) Possibly have something to do with the API being incorrect because of launching in development mode ?

Once again awesome work, will update post if i find anything else.

**Comment by Qix-** [comment]
Glad to hear it. 

Tests failing are also down from ~170 or so to just 37. I'm going through and revising how we handle paths but it'll take a bit of work - however I believe that's the source of most of the remaining failures.

> Git push fails.

Yeah the push/pull stuff is going to be a pain to fix, libgit2 famously doesn't do push/pull on Windows very well. I haven't even begun investigating that yet but will get to it at some point.

For now, you can do pushes manually via the CLI. All of our refs are stored under `.git/refs/gitbutler`, so you should be able to do `git push <remote> refs/gitbutler/<virtual branch name>` where `<virtual branch name>` is one of the list items from e.g. `ls .git/refs/gitbutler`. They should correspond to your virtual branch names.

> Images still fail to load.

This should be easy enough to fix; I believe @mtsgrd mentioned he might just do this from the frontend if possible instead of proxying through Tauri. We still need to investigate this one, as we've had issues with this even on other platforms in the past.

> Git commits are credited to GitButler regardless of setting that controls who is the author.

Very strange. Can you run `git config gitbutler.utmostDiscretion` in your repository and check that it's `1` or `true` or something? Make sure to have disabled committing with GitButler in the settings first.

> Can not sign in, get an error saying there was an issue, error in DevTools shows (Unauthorized) Possibly have something to do with the API being incorrect because of launching in development mode ?

Indeed, this is because you're under Dev mode. Barring other windows-related issues, this should fix itself once we cut official builds. At some point we'll probably try to get this working under dev mode but that's low priority for us at the moment.

Thanks again for all the feedback, keep it coming if you find any more :)

**Comment by Qix-** [comment]
#3262 just landed, adding fetch/push support for Windows via the "System Executable" auth method, settable via the project settings.

Please let me know how it goes :)

**Comment by DrBarnabus** [comment]
Been trying out a Windows build of GitBulter (from commit ce548aca) only issue I've really noticed is that the login refuses to work for me it's hitting a CORS error but interestingly only in nightly/release build not when running `pnpm tauri dev` there it works perfectly.

Great to see the Windows support actively being worked on!

CORS error for reference:
![Image](https://github.com/gitbutlerapp/gitbutler/assets/16451892/09aab134-4351-4b1c-9482-f327a468a648)

**Comment by Qix-** [comment]
Hey @DrBarnabus can you try again? We just updated our CORS policy to hopefully fix that. Shouldn't need to reinstall, perhaps just restart the app for good measure. Let me know :)

**Comment by Qix-** [comment]
A few updates:

- #3279 fixes profile images on windows.
- A number of PRs were just merged to get preliminary MSI builds working in CI. I'll work on NSIS packages soon to see if they're any better (MSI leaves a lot to be desired).

We're pretty close to having windows releases :) We're in the process of receives certificates to allow us (hopefully, if the Microsoft gods allow us) to not have the "Windows just protected your computer" dialogs pop up when you run the installer after downloading it. Once that's set up, we'll probably go ahead with an official release.

Thanks for bearing with us!

**Comment by DrBarnabus** [comment]
> Hey @DrBarnabus can you try again? We just updated our CORS policy to hopefully fix that. Shouldn't need to reinstall, perhaps just restart the app for good measure. Let me know :)
> 
> EDIT: Just heard back, it'll be about 30 minutes before it's live. Try again in about 30-45 minutes :)

Just retested and that's now working which is great. Saw nightly builds are available now I'll keep up to date and report any other issues!

**Comment by w5l** [comment]
That's great news! Apologies for asking the (probably_ obvious, but where can the nightlies be found?

**Comment by B1ack3ye** [comment]
> That's great news! Apologies for asking the (probably_ obvious, but where can the nightlies be found?

https://github.com/gitbutlerapp/gitbutler/tags

But there are no ready-made builds included. Only the repository at that time.

**Comment by mtsgrd** [comment]
You can find the nightlies on our downloads page! https://app.gitbutler.com/downloads

**Comment by A-Lamia** [comment]
Testing out the nightly, i can only move a single file to a virtual branch when i try to move a second i get the error `Failed to update branch ownership`.

**Comment by ahaoboy** [comment]
UI bug
![image](https://github.com/gitbutlerapp/gitbutler/assets/19884146/1681c7cd-8a63-4a74-b794-3abcb03da61b)

**Comment by w5l** [comment]
> You can find the nightlies on our downloads page! https://app.gitbutler.com/downloads

Thanks! I see that got a lot of attention and the first issues reported in this thread already. Shall we follow the [Bugs and Feature Requests](https://github.com/gitbutlerapp/gitbutler/blob/master/README.md#bugs-and-feature-requests) section of the main readme and make separate issue reports, or do you prefer issues dumped in this thread for now?

**Comment by Qix-** [comment]
Once we get the certs for publishing a non-nightly build and we can verify it's working with the folks here, I'll go ahead and close this issue and we can start with normal bug reports :) Still want to make sure we've not missed anything major, especially with the code signing, as it could cause people to have Windows Defender freak out if not done correctly.

**Comment by slingshotvfx** [comment]
I am unable to add a git repository from a mapped network drive on windows 11 nightly 0.5.379.0, it only pops up this message:

![image](https://github.com/gitbutlerapp/gitbutler/assets/146885925/95c7073d-12fb-4b29-8291-4e68ae785d2c)

I am able to add the repo if I copy it the local c: drive.  

Adding an existing repo gives me this message:
![image](https://github.com/gitbutlerapp/gitbutler/assets/146885925/e4c903af-a790-43cb-83d4-627b97745022)
And if I click "go back to gitbutler/integration":
![image](https://github.com/gitbutlerapp/gitbutler/assets/146885925/cfb01a9c-1aa5-453c-9342-70f2f903e671)


However, cloning it fresh from github to the c: drive and adding it works as expected.


I'd love to help test windows builds, are there debug logs somewhere or any other diagnostic information?

**Comment by Greedquest** [comment]
Note you have to right click + properties and Unblock the msi (at least on Win 11) to get it to run as it has the mark of the Web otherwise
![image](https://github.com/gitbutlerapp/gitbutler/assets/25348920/272229fa-54e4-4a53-bd21-0a1ae008ed06)

**Comment by Qix-** [comment]
Hey @slingshotvfx, I'm not sure of a workaround for this, this is, unfortunately, a huge limitation of the Windows network drives that affects a _lot_ of software. For example, Cargo won't even work for me if I `cd` into a network drive and try to build a Linux project from the Windows side.

Unless there's a trick that solves this somehow, it's not something I think we can work around :/ We should definitely document this, though - I think it should be trivial enough to determine if it's a network drive on Windows and show a warning/error.

**Comment by slingshotvfx** [comment]
@Qix- just to clarify, this was running the windows nightly and trying to access a repo on a windows drive (Z:\ in my case) that is mapped to an smb network share.  No WSL or cross-platform access shenanigans.

Adding a project from a UNC path (e.g. \\server\path\to\repo) also gives me the same error.

**Comment by Qix-** [comment]
Ah sorry, yeah I misread. It must just be a network share thing, unfortunately. I'll have to look a bit more into this to see if there's some sort of workaround, but I can't make any promises it'll happen soon.

**Comment by pmario** [comment]
@slingshotvfx 

>I am unable to add a git repository from a mapped network drive on windows 11 nightly 0.5.379.0, it only pops up this message:

I did have the same problem. I could fix it by installing the latest version of git. I did `git update`

**Comment by slingshotvfx** [comment]
> I did have the same problem. I could fix it by installing the latest version of git. I did `git update`

Unfortunately I still receive the same error with the latest version of git for windows.

**Comment by ccpwcn** [comment]
Windows system too many users, must support ah, come quickly, can't wait.

**Comment by B1ack3ye** [comment]
Is there any way to change the email address that git uses to push to specific projects?

I added the email to the global .gitconfig, but it still tries to push with a different email. No matter what "Git authentication" setting I configure in Gitbutler.
![image](https://github.com/gitbutlerapp/gitbutler/assets/63259739/a34216a4-63c0-4cb1-a4b5-34320edbf0ae)

**Comment by DrBarnabus** [comment]
Unsure if it's isolated just to windows, I'd imagine it is though as I couldn't see any similar reports in issues, but I ran into an error updating the base branch yesterday.

Raised https://github.com/gitbutlerapp/gitbutler/issues/3412

**Comment by 2Friendly4You** [comment]
Found two things on "Windows Nightly 0.5.400"

I'm not sure, if it should say OpenAI API, if I select Anthropic? Is that right?
<img width="446" alt="Screenshot 2024-04-08 204117" src="https://github.com/gitbutlerapp/gitbutler/assets/83555863/5fbf4f42-da08-4407-a0b6-8556a2c5b6db">


And one small, unimportant thing:
When selecting a profile picture for the personal profile, it would be practical if the file selection only displayed the supported picture types and not all file types.

**Comment by Qix-** [comment]
Hi all, we just launched signed Windows builds - check out https://gitbutler.com for them ([direct link to MSI](https://app.gitbutler.com/downloads/release/windows/x86_64/msi)).

Going to close this issue now; if there are any outstanding issues you're facing, **please open them as new issues** and let us know you're running under Windows so we can properly track them.

Keep in mind we're still in beta, so keep the bug reports flowing so we can address them and make things a bit less beta 🙃

Finally,

# Thank you to everyone who provided feedback ❤️ 

\- GitButler Team
